catalog/app/containers/Bucket/requests/packageUpload.ts

```diff
@@ -0,0 +1,258 @@
+import * as APIConnector from 'utils/APIConnector'
+import { makeSchemaDefaultsSetter, JsonSchema } from 'utils/json-schema/json-schema'
+import pipeThru from 'utils/pipeThru'
+import * as workflows from 'utils/workflows'
+
+// "CREATE package" - creates package from scratch to new target
+// "UPDATE package" - creates package from source (existing manifest) to the same target
+// "COPY package" - creates package from source (existing manifest) to new target
+// "WRAP package" - creates package (wraps directory) from source (bucket + directory) to new target
+
+interface FileEntry {
+  is_dir: boolean
+  logical_key: string
+  path: string
+}
+
+interface FileUpload {
+  hash: string
+  logical_key: string
+  physical_key: {
+    bucket: string
+    key: string
+    version: string
+  }
+  size: number
+}
+
+interface ManifestBodyBase {
+  message: string
+  meta: JsonValue
+  registry: string
+  workflow?: string | null
+}
+
+interface ManifestBodyCreate extends ManifestBodyBase {
+  contents: FileUpload[]
+  name: string
+}
+
+interface ManifestBodyCopy extends ManifestBodyBase {
+  name: string
+  parent: {
+    top_hash: string
+    registry: string
+    name: string
+  }
+}
+
+interface ManifestBodyWrap extends ManifestBodyBase {
+  dst: {
+    registry: string
+    name: string
+  }
+  entries: FileEntry[]
+}
+
+type ManifestBody = ManifestBodyCreate | ManifestBodyWrap | ManifestBodyCopy
+
+const ENDPOINT_CREATE = '/packages'
+
+const ENDPOINT_UPDATE = '/packages'
+
+const ENDPOINT_COPY = '/packages/promote'
+
+const ENDPOINT_WRAP = '/packages/from-folder'
+
+type Endpoint =
+  | typeof ENDPOINT_CREATE
+  | typeof ENDPOINT_UPDATE
+  | typeof ENDPOINT_COPY
+  | typeof ENDPOINT_WRAP
+
+// TODO: reuse it from some other place, don't remember where I saw it
+interface ManifestHandleTarget {
+  bucket: string
+  name: string
+}
+
+interface ManifestHandleSource extends ManifestHandleTarget {
+  revision: string
+}
+
+interface BasePackageParams {
+  message: string
+  meta: JsonValue
+  workflow: workflows.Workflow
+}
+
+interface CreatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  target: {
+    bucket: string
+    name: string
+  }
+}
+
+interface UpdatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  source: {
+    bucket: string
+    name: string
+  }
+}
+
+interface CopyPackageParams extends BasePackageParams {
+  source: ManifestHandleSource
+  target: ManifestHandleTarget
+}
+
+interface WrapPackageParams extends BasePackageParams {
+  entries: FileEntry[]
+  source: string
+  target: ManifestHandleTarget
+}
+
+interface Response {
+  top_hash: string
+}
+
+// FIXME: this is copypasted from PackageDialog -- next time we need to TSify utils/APIConnector properly
+interface ApiRequest {
+  <O>(opts: {
+    endpoint: string
+    method?: 'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD'
+    body?: {}
+  }): Promise<O>
+}
+
+const uploadManifest = (
```

Comment:
i dont think this function should exist
1) its name doesnt reflect all its use cases, it looks like a leaked abstraction;
2) it doesnt have enough logic, i think it's better to just call `req` directly (you can parameterize body and response types to add type checking there)

Reply:
It's a place where I can validate the request body with Schema

Reply:
you can validate it the same way by using type arg for `req`